packages/server-wallet/src/onchain-service/__chain-test__/transaction-submission.test.ts

```diff
@@ -0,0 +1,111 @@
+    value: BigNumber.from(10),
+    data: '0x00',
+    ...overrides,
+  };
+};
+
+const getFailingWallet = (error?: string): any => {
+  return {
+    connect: (_provider: providers.JsonRpcProvider): any => {
+      return {
+        getTransactionCount: (): number => 0,
+        sendTransaction: (
+          _tx: providers.TransactionRequest
+        ): Promise<providers.TransactionResponse> => {
+          throw new Error(error || 'Fail');
+        },
+      };
+    },
+  };
+};
+
+describe('TransactionSubmissionService.submitTransaction', () => {
+  let service: TransactionSubmissionService;
+  let wallet: Wallet;
+  let provider: providers.JsonRpcProvider;
+
+  const channelId = randomChannelId();
+
+  beforeEach(async () => {
+    provider = new providers.JsonRpcProvider(config.rpcEndpoint);
+    wallet = new Wallet(config.serverPrivateKey);
+    service = new TransactionSubmissionService(provider, wallet, new TransactionSubmissionStore());
+  });
+
+  it('should fail if provided with 0 attempts', async () => {
+    const tx = getTransaction();
+    await expect(service.submitTransaction(channelId, tx, {maxSendAttempts: 0})).rejects.toEqual(
+      new TransactionSubmissionError(TransactionSubmissionError.reasons.zeroAttempts)
+    );
+  });
+
+  it('should send a transaction successfully', async () => {
+    const tx = getTransaction();
+    const response = await service.submitTransaction(channelId, tx);
+    await verifyResponse(response, tx);
+  });
+
+  it('should send concurrent transactions successfully', async () => {
+    const tx = getTransaction();
+    const responses = await Promise.all([
+      service.submitTransaction(channelId, tx),
+      service.submitTransaction(channelId, tx),
+    ]);
+    await Promise.all(responses.map(response => verifyResponse(response, tx)));
+  });
+
+  it('should retry transactions with known errors', async () => {
+    // Create a "wallet" that forces an error
+    const mock = getFailingWallet(TransactionSubmissionError.knownErrors.badNonce);
+
+    // Recreate service
+    service = new TransactionSubmissionService(
+      provider,
+      mock as any,
+      new TransactionSubmissionStore()
+    );
+
+    const tx = getTransaction();
+    await expect(service.submitTransaction(channelId, tx, {maxSendAttempts: 2})).rejects.toEqual(
+      new TransactionSubmissionError(TransactionSubmissionError.reasons.failedAllAttempts)
+    );
+  });
+
+  it('should not retry transactions with unknown errors', async () => {
+    // Create a "wallet" that forces an unrecognized error
+    const mock = getFailingWallet();
+
+    // Recreate service
+    service = new TransactionSubmissionService(provider, mock, new TransactionSubmissionStore());
+
+    const tx = getTransaction();
+    await expect(service.submitTransaction(channelId, tx, {maxSendAttempts: 2})).rejects.toEqual(
+      new TransactionSubmissionError(TransactionSubmissionError.reasons.unknownError)
+    );
+  });
+});
```

Comment:
This comment is for the [`PNG` file](https://github.com/statechannels/statechannels/blob/5d5b5947cf75e7e32f5c962763fa9209bc74c354/packages/server-wallet/src/onchain-service/assets/onchain-architecture.png) that I see below. I can't comment on it directly.

I'm not sure about including that architecture diagram here, because
A. it is quite different from the way the browser-wallet works
B. it is quite different from the way the server-wallet will work (which is a bit different from A)
C. it uses different terminology than the rest of this codebase